### PR TITLE
[jit] add shallow copy to orderdictwrapper

### LINKED ISSUE #15421: JIT is not compatible with data parallel
## 🐛 Bug

JIT is not compatible with data parallel. So no one can train any model using more than one GPU.

## To Reproduce

1. Use resnet 18 script module from https://github.com/pytorch/pytorch/blob/0b219538cfb70503baa74053d3e827951e0ea6f9/test/test_jit.py#L5882-L5983
1. `model = torch.nn.DataParallel(model)`
1. Get the following error

> Traceback (most recent call last):
>   File "train.py", line 459, in <module>
>     main()
>   File "train.py", line 333, in main
>     train(train_loader, model, criterion, optimizer, epoch)
>   File "train.py", line 394, in train
>     output = model(input)
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 489, in __call__
>     result = self.forward(*input, **kwargs)
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 142, in forward
>     replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 147, in replicate
>     return replicate(module, device_ids)
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/replicate.py", line 31, in replicate
>     replica._parameters = replica._parameters.copy()
> AttributeError: 'OrderedParameterDict' object has no attribute 'copy'

If I firstly run `model = torch.nn.DataParallel(model)` for a nn.Module, then trace it, although the code runs, the loss doesn't decrease.

## Expected behavior

The code runs and the model gets trained

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla P40
GPU 1: Tesla P40
GPU 2: Tesla P40

Nvidia driver version: 390.77
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

## Additional context

Related issue: #12875 

Data parallel is such a fundamental module, I can't imagine JIT is not compatible with it and no official programmer have tested it. It makes the JIT totally unpractical for training (except toy models)

### PR DESCRIPTION
**Pull request by wanchaol**
Fixes #15421

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -813,6 +814,9 @@ def __getitem__(self, k):
     def __setitem__(self, k, v):
         raise NotImplementedError
 
+    def copy(self):
+        return self.__class__(self)
```
Comment (driazati) [inline torch/jit/__init__.py]: Shouldn't this be `self.__class__(self.submodule.get())`?

**Comment by daquexian** [comment]
Thanks for trying solve the problem! I tried it and got the following error:

```
 File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 147, in replicate
    return replicate(module, device_ids)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/replicate.py", line 32, in replicate
    replica._buffers = replica._buffers.copy()
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/__init__.py", line 1328, in __setattr__
    return super(TracedModule, self).__setattr__(attr, value)
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/__init__.py", line 1119, in __setattr__
    return super(ScriptModule, self).__setattr__(attr, value)
  File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 550, in __setattr__
    elif params is not None and name in params:
  File "/usr/local/lib/python3.6/dist-packages/torch/jit/__init__.py", line 864, in __contains__
    return self.module._has_parameter(k)
AttributeError: 'OrderedParameterDict' object has no attribute '_has_parameter'
```

File: torch/jit/__init__.py (hunk 1 above)
Comment (daquexian) [reply torch/jit/__init__.py]: > Shouldn't this be `self.__class__(self.submodule.get())`?

@driazati Thanks for your comment. I'm the author of #15421 . I tried `self.__class__(self.submodule.get())` but got `AttributeError: 'OrderedParameterDict' object has no attribute 'submodule'`

**Comment by daquexian** [comment]
@wanchaol @driazati Could you please tell me is there any progress? :)

**Comment by daquexian** [comment]
I tried to solve this issue today but failed again.
After change
```
def copy(self):
        return self.__class__(self)
```
to
```
def copy(self):
        return self.__class__(self.module())
```
I met a new error.

>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py", line 489, in __call__
>     result = self.forward(*input, **kwargs)
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 142, in forward
>     replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/data_parallel.py", line 147, in replicate
>     return replicate(module, device_ids)
>   File "/usr/local/lib/python3.6/dist-packages/torch/nn/parallel/replicate.py", line 46, in replicate
>     replica._modules[key] = module_copies[j][module_idx]
>   File "/usr/local/lib/python3.6/dist-packages/torch/jit/__init__.py", line 840, in __setitem__
>     self.module._register_module(k, v)
> RuntimeError: Unable to cast from non-held to held instance (T& to Holder<T>) of type 'std::shared_ptr<torch::jit::script::Module>''

I'm not familiar to pybind11, so I just jumped over this error by changing 
```
replica._modules[key] = module_copies[j][module_idx]
```
to
```
replica._modules[key] = ScriptModule()
```
And then I got
RuntimeError: Module 'conv1' already defined (insert at /pytorch/torch/csrc/api/include/torch/ordered_dict.h:359)
So obviously, we cannot create the dicts by the same original module.

I tried to create the dicts by the new module but got segfault.

These are all I found before I finally gave up.

**Comment by daquexian** [comment]
I'm surprised for the following three things

1. Facebook announced **pytorch 1.0 stable** proudly and leave such a fundamental thing unimplemented.
2. An official programmer confidently posted his pr without running his code at least once. (The solution is much more than adding a simple `copy` method)
3. No one cares about it in nearly a week.

What a shame

**Comment by soumith** [comment]
@daquexian we are sorry, this is the holidays week, so quite a few people are on vacation, hence the delay in responses.

**Comment by zdevito** [comment]
I looked into the requirements for this fix a bit more and it will require use to develop a more complete solution to replicating modules than what replicate.py already has. In the meantime a possible workaround is to use DistributedDataParallel, which I believe avoids having to perform the replicate in the first place.

In particular we need to define a overridable protocol for copying a module instance and then make that work for ScriptModule and its subclasses.  It will take some time to sort out how to do that for (Weak)ScriptModule because there is a lot going on in those classes (they derive from pybind11 objects, they have a custom meta-class, the initializers actually have to run for the class to be valid, etc.).

**Comment by daquexian** [comment]
@soumith sorry for my comment on impulse. I'm willing to provide information to help solve this issue if you want.
@zdevito thanks for your work. I will try DistributedDataParallel.

**Comment by soumith** [comment]
@daquexian no worries, it is understandable, we did not respond for several days, and it's not holidays everywhere in the world, so you probably didn't know what was going on our side :)

DistributedDataParallel is your best solution, because it's a cleaner design that just spawns multiple python processes, each process controlling 1 GPU for example. Our imagenet example has example usage of DistributedDataParallel here: https://github.com/pytorch/examples/tree/master/imagenet#multi-processing-distributed-data-parallel-training

About why we did not think of `nn.DataParallel + ScriptModule` for 1.0 stable, it is because we mainly focused on exportability / production aspects for 1.0, it is a bit of a blind-spot on our side, sorry about that.

**Comment by daquexian** [comment]
@soumith Thanks, I tried ddp successfully, though I didn't observe speedup obtain by jit (maybe I did something wrong, I'm a little tired to figure it out)

BTW I fixed a bug of the training script caused by a recent commit (https://github.com/pytorch/examples/pull/484)

**Comment by wanchaol** [comment]
@daquexian sorry for the late response, I was traveling for the past few days and did not get access to a stable internet till today. I sketched this patch and it worked on a simple ScriptModule initially, but later I found it fails to scale to resnet because of the complications in replicate.py, sorry for not updating on the PR to let you know. Thanks @soumith @zdevito on help resolve this and provide an alternative way to get around it, I will just close this PR as Zach mentioned we will have a more complete solution on this.

**Pull request closed without merging**